src/custom-elements/src/components/number-input/number-input.tsx

```diff
@@ -84,114 +112,98 @@ export class NumberInput {
+    this.chiInput.emit(this.value);
 
-    if (newValue !== this.value) {
-      this._setNewValue(newValue);
-    } else {
-      if ((ev.target as HTMLInputElement).value.toString() !== this.value) {
-        (ev.target as HTMLInputElement).value = this.value;
-      }
-    }
+    if (!input.validity.valid) this.chiNumberInvalid.emit();
   }
 
   private increment() {
-    const step = this.step;
-    const newValue = Math.round((+this.value + step) * 10000) / 10000;
+    const input = this.el.id
+      ? (document.getElementById(`${this.el.id}-control`) as HTMLInputElement)
+      : (this.el.children[0].children[0] as HTMLInputElement);
+
+    input.stepUp();
+    this.value = this.setNewValue(input.value);
 
-    if (newValue <= this.max) {
-      this.value = newValue.toString();
+    if (input.valueAsNumber <= this.max) {
       this._didUpdateCallBackOnceQueue.push(() => {
         this.chiChange.emit(this.value);
       });
     }
   }
 
   private decrement() {
-    const step = this.step;
-    const newValue = Math.round((+this.value - step) * 10000) / 10000;
+    const input = this.el.id
+      ? (document.getElementById(`${this.el.id}-control`) as HTMLInputElement)
+      : (this.el.children[0].children[0] as HTMLInputElement);
 
-    if (newValue >= this.min) {
-      this.value = newValue.toString();
+    input.stepDown();
+    this.value = this.setNewValue(input.value);
+
+    if (input.valueAsNumber >= this.min) {
       this._didUpdateCallBackOnceQueue.push(() => {
         this.chiChange.emit(this.value);
       });
     }
   }
 
+  private formatNumber(number: string): string {
+    return parseFloat(number).toFixed(this.decimals);
+  }
+
+  private setNewValue(number: string): string {
+    return this.decimals ? this.formatNumber(number) : number;
+  }
+
   render() {
     const input = (
       <input
         type="number"
         class={`chi-input ${this.inputstyle ? `-${this.inputstyle}` : ''} ${
-          this.state ? `-${this.state}` : ''}`}
+          this.state ? `-${this.state}` : ''
+        }`}
         disabled={this.disabled}
         required={this.required}
         step={this.step}
         max={this.max}
         min={this.min}
         value={this.value}
         onChange={ev => this.handleChange(ev)}
+        onInput={ev => this.handleInput(ev)}
         id={this.el.id ? `${this.el.id}-control` : null}
       />
     );
 
     const base = (
-      <div
-        class={`chi-number-input ${this.size ? `-${this.size}` : ''}`}
-      >
+      <div class={`chi-number-input ${this.size ? `-${this.size}` : ''}`}>
         {input}
         <button
-          disabled={+this.value - this.step < this.min}
+          disabled={+this.value <= this.min}
           onClick={() => this.decrement()}
           aria-label="Decrease"
         ></button>
         <button
-          disabled={+this.value + this.step > this.max}
+          disabled={
+            +this.value + this.step > this.max || +this.value >= this.max
```

Comment:
Why is the first `+` character necessary here?

Reply:
`this.value` is a string attribute, so the `+` operator converts it to a number to perform the operation.